File: codeQuest_2025/countingCards.py
for i in range(int(input())):
    playerHand = input().split(" ")
    dealerHand = input().split(" ")
    playerScore = 0
    playerAces = 0
    dealerScore = 0
    dealerAces = 0
    key = {
        "2": 2,
        "3": 3,
        "4": 4,
        "5": 5,
        "6": 6,
        "7": 7,
        "8": 8,
        "9": 9,
        "10": 10,
        "JACK": 10,
        "QUEEN": 10,
        "KING": 10,
        "ACE": 11  # Start Aces as 11
    }

    # Calculate player score
    for item in playerHand:
        splitItem = item.split("_")
        card = splitItem[0]
        
        if card == "ACE":
            playerAces += 1
            playerScore += key[card]  # Add Ace as 11
        else:
            playerScore += key[card]
    
    # Adjust for Aces if score exceeds 21
    while playerScore > 21 and playerAces > 0:
        playerScore -= 10  # Adjust one Ace from 11 to 1
        playerAces -= 1

    # Calculate dealer score
    for item in dealerHand:
        splitItem = item.split("_")
        card = splitItem[0]
        
        if card == "ACE":
            dealerAces += 1
            dealerScore += key[card]  # Add Ace as 11
        else:
            dealerScore += key[card]

    # Adjust for Aces if score exceeds 21
    while dealerScore > 21 and dealerAces > 0:
        dealerScore -= 10  # Adjust one Ace from 11 to 1
        dealerAces -= 1

    print("Player Score: " + str(playerScore) + " Dealer Score: " + str(dealerScore), end="")
    
    if playerScore > 21:
        print(" Dealer Wins!")
    elif dealerScore > 21 or playerScore > dealerScore:
        print(" Player Wins!")
    elif dealerScore > playerScore:
        print(" Dealer Wins!")
    else:
        print(" Tie!")

File: CodeQuest2024/countingCards.java
import java.util.*;
import java.lang.*;
import java.text.*;
public class MyProgram{
public static void main(String[] args) {
    try (Scanner input = new Scanner(System.in)) {
      int testCases = Integer.parseInt(input.nextLine());
      for (int c= 0; c< testCases; c++) {
        String line1 = input.nextLine();
        String line2 = input.nextLine();
        String[] playerArr = line1.split(" ");
        String[] dealerArr = line2.split(" ");
        int playerSum = 0;
        int playerAces = 0;
        int dealerSum = 0;
        int dealerAces = 0;
        for(int i = 0; i < playerArr.length; i++){
            String temp = playerArr[i].substring(0, playerArr[i].indexOf("_"));
            if(temp.equals("ACE")){
                playerAces++;
            }
            else if(temp.equals("KING") || temp.equals("QUEEN") || temp.equals("JACK")){
                playerSum += 10;
            }
            else{
                playerSum += Integer.valueOf(temp);
            }
        }
        for(int i = 0; i < dealerArr.length; i++){
            String temp = dealerArr[i].substring(0, dealerArr[i].indexOf("_"));
            if(temp.equals("ACE")){
                dealerAces++;
            }
            else if(temp.equals("KING") || temp.equals("QUEEN") || temp.equals("JACK")){
                dealerSum += 10;
            }
            else{
                dealerSum += Integer.valueOf(temp);
            }
        }
        //2 for loops determining if adding aces will bust or not,
        //and adding aces to score appropriately(if it will bust, 1)
        for(int i = 0; i < playerAces; i++){
            if(playerSum + 11 > 21){
                playerSum += 1;
            }
            else{
                playerSum += 11;
            }
        }
        for(int i = 0; i < dealerAces; i++){
            if(dealerSum + 11 > 21){
                dealerSum += 1;
            }
            else{
                dealerSum += 11;
            }
        }
        //then logic to see who wins  
        if((dealerSum == playerSum) || (dealerSum > 21 && playerSum > 21)){
            System.out.println("Player Score: " + playerSum + " Dealer Score: " + dealerSum +  " Tie!");
            continue;
        }
        if(dealerSum > 21){
            System.out.println("Player Score: " + playerSum + " Dealer Score: " + dealerSum +  " Player Wins!");
        }
        else if(playerSum > 21){
            System.out.println("Player Score: " + playerSum + " Dealer Score: " + dealerSum +  " Dealer Wins!");
        }
        else{
            if((21 - dealerSum) > (21 - playerSum)){
                System.out.println("Player Score: " + playerSum + " Dealer Score: " + dealerSum +  " Player Wins!");
            }
            else{
                System.out.println("Player Score: " + playerSum + " Dealer Score: " + dealerSum +  " Dealer Wins!");
            }
        }
      }
      }
    }
}